### ISSUE TITLE
Consider removing newlines between uncommented interface/struct members

### ISSUE THREAD

**Issue by flimzy** [mvdan/gofumpt#85]
I'm not sure if this is a reasonable addition, or if the heuristics need to be fine-tuned slightly to not be too aggressive, but my suggestion:

```golang
type Foo interface{

    Foo(context.Context) bool

    Bar(context.Context) (int64, error)

    Baz(contex.Context, string) error

}
```

becomes:

```golang
type Foo interface{
    Foo(context.Context) bool
    Bar(context.Context) (int64, error)
    Baz(contex.Context, string) error
}
```

But commented interface members can retain their whitespace, except at the very end/beginning of a delcaration:

```golang
type Foo struct{

    // Foo is very special
    Foo bool

    // Bar is dull.
    Bar int64

    Baz string

    Qux float64

    // Quxx is often misunderstood
    Quxx chan bool

}
```

```golang
type Foo struct{
    // Foo is very special
    Foo bool

    // Bar is dull.
    Bar int64

    Baz string
    Qux float64

    // Qux is often misunderstood
    Qux chan bool
}
```

Multi-line declarations may or may not deserve special consideration as well.

**Comment by mvdan** [comment]
I tend to agree that sometimes too many empty lines are used, such as in your first example. However, I'm worried about false positives. For example, it's perfectly reasonable to want to group fields by logical groups:

```
// First the exported ones, then the unexported ones.

type T {
    Foo bool
    Bar int64

    baz string
}
```

I've personally done similar groupings to clearly separate which fields are set when constructing T, and only read afterwards.

You could argue I wouldn't be bitten by this false positive if I documented the fields, but that would be surprising to many users, I bet.

**Comment by flimzy** [comment]
> I've personally done similar groupings to clearly separate which fields are set when constructing T, and only read afterwards.

:thinking: Does this, or similar reasoning, apply to interfaces?

**Comment by mvdan** [comment]
From my own experience, I think only structs. Also because in structs, the order actually matters for alignment and memory layout. There are also many common idioms, like:

```
[other fields]

fooMu sync.Mutex
foo T2

[other fields]
```

If most of your pain comes from interfaces, I think it would be reasonable to do it there. It should be easy enough to do, if you'd want to have a crack at it.

**Comment by Zamiell** [comment]
> I tend to agree that sometimes too many empty lines are used, such as in your first example. However, I'm worried about false positives. For example, it's perfectly reasonable to want to group fields by logical groups

I just want to echo this. I very commonly break up sections in a big struct with a newline to separate related fields. And I don't bother using a comment when I do this, since it is obvious what I am doing, and the comment would be superfluous:

```go
type myBigStruct struct {
    fooThing string
    fooOtherThing int
    fooAnotherThing []string

    barThing string
    barOtherThing int
    barAnotherThing []string
}
```

**Comment by mvdan** [comment]
Fixed by 835205518b7ab575a65de6148d29577dac46050d.
